packages/material-ui/src/Avatar/Avatar.js

```diff
@@ -9,15 +10,12 @@ import avatarClasses, { getAvatarUtilityClass } from './avatarClasses';
 const overridesResolver = (props, styles) => {
   const { variant = 'circular' } = props;
 
-  const styleOverrides = {
-    ...styles.root,
+  return deepmerge(styles.root || {}, {
```

Comment:
Is there a performance overhead at using deep merge? (runtime)

On are related note:

1. It seems that we never need the name

```diff
diff --git a/packages/material-ui/src/styles/experimentalStyled.js b/packages/material-ui/src/styles/experimentalStyled.js
index 400865ef9e..a82bd042a9 100644
--- a/packages/material-ui/src/styles/experimentalStyled.js
+++ b/packages/material-ui/src/styles/experimentalStyled.js
@@ -105,7 +105,7 @@ const experimentalStyled = (tag, options, muiOptions = {}) => {
     if (name && muiOptions.overridesResolver) {
       expressionsWithDefaultTheme.push((props) => {
         const theme = isEmpty(props.theme) ? defaultTheme : props.theme;
-        return muiOptions.overridesResolver(props, getStyleOverrides(name, theme), name);
+        return muiOptions.overridesResolver(props, getStyleOverrides(name, theme));
       });
     }
```

2. `theme` seems to always be defined, no need to test it

```diff
diff --git a/packages/material-ui/src/styles/experimentalStyled.js b/packages/material-ui/src/styles/experimentalStyled.js
index 400865ef9e..7e55aa5570 100644
--- a/packages/material-ui/src/styles/experimentalStyled.js
+++ b/packages/material-ui/src/styles/experimentalStyled.js
@@ -11,7 +11,6 @@ const getStyleOverrides = (name, theme) => {
   let styleOverrides = {};

   if (
-    theme &&
     theme.components &&
     theme.components[name] &&
     theme.components[name].styleOverrides
```

3. How about we add an early exit to improve the runtime?

```diff
diff --git a/packages/material-ui/src/styles/experimentalStyled.js b/packages/material-ui/src/styles/experimentalStyled.js
index 400865ef9e..3c1f93cfe9 100644
--- a/packages/material-ui/src/styles/experimentalStyled.js
+++ b/packages/material-ui/src/styles/experimentalStyled.js
@@ -8,18 +8,15 @@ function isEmpty(obj) {
 }

 const getStyleOverrides = (name, theme) => {
-  let styleOverrides = {};
-
   if (
     theme.components &&
     theme.components[name] &&
     theme.components[name].styleOverrides
   ) {
-    styleOverrides = theme.components[name].styleOverrides;
+    return theme.components[name].styleOverrides;
   }
-
-  return styleOverrides;
+  return null;
 };

 const getVariantStyles = (name, theme) => {
@@ -105,7 +102,13 @@ const experimentalStyled = (tag, options, muiOptions = {}) => {
     if (name && muiOptions.overridesResolver) {
       expressionsWithDefaultTheme.push((props) => {
         const theme = isEmpty(props.theme) ? defaultTheme : props.theme;
-        return muiOptions.overridesResolver(props, getStyleOverrides(name, theme), name);
+        const styleOverrides = getStyleOverrides(name, theme);
+
+        if (styleOverrides) {
+          return muiOptions.overridesResolver(props, styleOverrides);
+        }
+
+        return null;
       });
     }
```

4. It seems that we forgot to rename the name for the box:

```diff
diff --git a/packages/material-ui/src/Box/Box.js b/packages/material-ui/src/Box/Box.js
index 3fe0153dc8..0a24b2578f 100644
--- a/packages/material-ui/src/Box/Box.js
+++ b/packages/material-ui/src/Box/Box.js
@@ -58,4 +58,4 @@ Box.propTypes = {
   sx: PropTypes.object,
 };

-export default styled(Box, {}, { muiName: 'MuiBox' })``;
+export default styled(Box, {}, { name: 'Box' })``;
```

However, shouldn't we add an option to skip the variant resolver for performance reasons with the box?

5. The current API for styled prevents its usage for custom components. Shouldn't we include Mui in the name? e.g.

```diff
diff --git a/packages/material-ui/src/Avatar/Avatar.js b/packages/material-ui/src/Avatar/Avatar.js
index 19cd9fcdcb..ea7c60dc02 100644
--- a/packages/material-ui/src/Avatar/Avatar.js
+++ b/packages/material-ui/src/Avatar/Avatar.js
@@ -38,7 +38,7 @@ const AvatarRoot = experimentalStyled(
   'div',
   {},
   {
-    name: 'Avatar',
+    name: 'MuiAvatar',
     slot: 'Root',
     overridesResolver,
   },
@@ -75,7 +75,7 @@ const AvatarImg = experimentalStyled(
   'img',
   {},
   {
-    name: 'Avatar',
+    name: 'MuiAvatar',
     slot: 'Img',
   },
 )({
@@ -94,7 +94,7 @@ const AvatarFallback = experimentalStyled(
   Person,
   {},
   {
-    name: 'Avatar',
+    name: 'MuiAvatar',
     slot: 'Fallback',
   },
 )({
diff --git a/packages/material-ui/src/Box/Box.js b/packages/material-ui/src/Box/Box.js
index 3fe0153dc8..0a24b2578f 100644
--- a/packages/material-ui/src/Box/Box.js
+++ b/packages/material-ui/src/Box/Box.js
@@ -58,4 +58,4 @@ Box.propTypes = {
   sx: PropTypes.object,
 };

-export default styled(Box, {}, { name: 'Box' })``;
+export default styled(Box, {}, { name: 'MuiBox' })``;
diff --git a/packages/material-ui/src/styles/experimentalStyled.js b/packages/material-ui/src/styles/experimentalStyled.js
index 400865ef9e..575015e504 100644
--- a/packages/material-ui/src/styles/experimentalStyled.js
+++ b/packages/material-ui/src/styles/experimentalStyled.js
@@ -77,8 +77,8 @@ const experimentalStyled = (tag, options, muiOptions = {}) => {

   if (componentName) {
     displayName = `${componentName}${componentSlot || ''}`;
-    name = !componentSlot || componentSlot === 'Root' ? `Mui${componentName}` : null;
-    className = `Mui${componentName}-${lowercaseFirstLetter(componentSlot || 'Root')}`;
+    name = !componentSlot || componentSlot === 'Root' ? `${componentName}` : null;
+    className = `${componentName}-${lowercaseFirstLetter(componentSlot || 'Root')}`;
   }

   const defaultStyledResolver = styled(tag, {
```

Reply:
Address all recommendations and updated the tests. I don't think we have suitable tests for benchmark the theme overrides so far. On a related note on the perf, maybe we could do the `deepmerge` only if the `styles.root` contains some of the class selectors for the other slots. This should definitely at least help us do the `deepmerge` only when really necessary. Should I add that check on all components?

Reply:
I think that we can stop here. "make it work, make it right, make it fast". Without a benchmark, I have no idea if these checks couldn't be worse. 

I was hopping that:

```diff
diff --git a/packages/material-ui/src/Avatar/Avatar.js b/packages/material-ui/src/Avatar/Avatar.js
index 84c332e478..0dbad464a4 100644
--- a/packages/material-ui/src/Avatar/Avatar.js
+++ b/packages/material-ui/src/Avatar/Avatar.js
@@ -10,11 +10,11 @@ const overridesResolver = (props, styles) => {
   const { variant = 'circular' } = props;

   return {
-    ...styles.root,
-    ...styles[variant],
     [`&.${avatarClasses.colorDefault}`]: styles.colorDefault,
     [`&.${avatarClasses.img}`]: styles.img,
     [`&.${avatarClasses.fallback}`]: styles.fallback,
+    ...styles.root,
+    ...styles[variant],
   };
 };
```

would be enough. It would force developers to choose between the two possible approaches and not mix them but why not:

OK

```jsx
components: {
  MuiAvatar: {
    styleOverrides: {
      root: {
        color: 'blue';
        ['& .slotClassName']: {
          color: 'red',
        },
      },
    },
  },
},
```

OK
```jsx
components: {
  MuiAvatar: {
    styleOverrides: {
      root: {
        color: 'blue';
      },
      slotClassName: {
        color: 'red',
      },
    },
  },
},
```

Raise a warning:

```jsx
components: {
  MuiAvatar: {
    styleOverrides: {
      root: {
        color: 'blue';
        ['& .slotClassName']: {
          color: 'red',
        },
      },
      slotClassName: {
        color: 'yellow',
      },
    },
  },
},
```

Reply:
If this really becomes a perf issue, we can definitely go with the warning in the future, at least production perf would not be affected by it 👍 Will leave the comment open for future reference